[git-webkit] Allow user to CC specific radar

JonWBedard opened the pull request:
#### d3db0af985b0d0b7c6d070b90934db4f901c1357
<pre>
[git-webkit] Allow user to CC specific radar
<a href="https://bugs.webkit.org/show_bug.cgi?id=245202">https://bugs.webkit.org/show_bug.cgi?id=245202</a>
&lt;rdar://problem/99942952&gt;

Unreviewed follow-up fix.

* Tools/Scripts/libraries/webkitbugspy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.cc_radar): Use radar number instead of bugzilla number.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/branch_unittest.py:
(TestBranch.test_manual_radar_cc): Correct test baseline.
(TestBranch.test_manual_radar_cc_integer): Ditto.
</pre>

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -138,11 +138,25 @@ def main(cls, args, repository, why=None, redact=False, **kwargs):
-                    break
+        needs_radar = issue and not isinstance(issue.tracker, radar.Tracker)
+        needs_radar = needs_radar and any([
+            isinstance(tracker, radar.Tracker) and tracker.radarclient()
```
Just to check my understanding, `radarclient()` is None on external contributors’ machines, so we won’t start prompting non-Apple folks to CC a Radar on every bug. Is that correct?

JonWBedard replied:
Correct. `radarclient()` is only available to Apple Internal folks.

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -138,11 +138,25 @@ def main(cls, args, repository, why=None, redact=False, **kwargs):
+            if not getattr(args, 'defaults', None):
+                sys.stdout.write('Existing radar to CC (leave empty to create new radar)')
+                sys.stdout.flush()
+                rdar = Tracker.from_string(Terminal.input(': '))
```
What formats does this recognize? NNNN, rdar://NNNN, rdar://problem/NNNN, \<rdar://NNNN\> \<rdar://problem/NNNN\> are all the ways I can think of. It’d be nice to support them all for easy copy-paste.

Relatedly, what happens if the input string does not parse as a radar URL?

JonWBedard replied:
Everything except for the first one will work here, and `Tracker.from_string` will return `None` if it doesn't match. I can make the `NNNN` approach work, the only reason it doesn't is because it's ambiguous and might not refer to radar, in general, but that's not true in this particular piece of code.

JonWBedard replied:
Following up more specifically, [webkitbugspy/radar.py](https://github.com/WebKit/WebKit/blob/a6840766432adbc9769019526c502c99476e8696/Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py#L42) shows you what we recognize as a radar string.

emw-apple replied:
Seems reasonable. I think I'm a little burned from back when we were using `webkit-patch`, and that _only_ accepted a bug number sans URL. But that memory will fade away :)

emw-apple commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py:
```diff
@@ -155,11 +155,10 @@ def _issue(self, url, id, credentials=None, data=None):
                     continue
                 radar_id = issue['id']
                 if RadarMock.top:
-                    print('Ugh, this is the problem')
```
😂

JonWBedard replied:
Yeah, testing code I forgot to delete from a pervious change :)

emw-apple commented on Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
```diff
@@ -479,28 +479,52 @@ def cc_radar(self, issue, block=False, timeout=None):
+            if user_to_cc:
+                keyword_to_add = 'InRadar'
+            elif comment_to_make:
+                sys.stderr.write('{} already CCed and tracking a different bug\n'.format(self.radar_importer.name))
```
Could we print the different bug being tracked here?

JonWBedard replied:
We could, yes.

emw-apple reviewed (approved):
While you’re doing this, could we also make `git-webkit pr` parse a radar URL from the commit message, and CC it if needed? For a commit message like:

    Commit subject line
    https://bugs.webkit.org/show_bug.cgi?id=NNNNNN
    rdar://NNNNNN

The second line already gets parsed out and commented on, so do the same for the third line. This would be a better workflow for people who create branches and commits manually, before running git-webkit.

JonWBedard committed:
[git-webkit] Allow user to CC specific radar
https://bugs.webkit.org/show_bug.cgi?id=245202
<rdar://problem/99942952>

Reviewed by Elliott Williams.

* Tools/Scripts/libraries/webkitbugspy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/bugzilla.py:
(Tracker.populate): Allow related issue to have the same number as self if trackers are different types.
(Tracker.cc_radar): Allow caller to specify radar to CC.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/github.py:
(Issue.cc_radar):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/issue.py:
(Issue.cc_radar):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/mocks/bugzilla.py:
(Bugzilla._issue): Skip creating radar if 'InRadar' keyword is present.
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/radar.py:
(Tracker.cc_radar):
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tests/bugzilla_unittest.py:
* Tools/Scripts/libraries/webkitbugspy/webkitbugspy/tracker.py:
(Tracker.cc_radar):
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
(Branch.main): Ask the user to provide an existing radar instead of always creating a new one.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/branch_unittest.py:
(TestBranch.test_automatic_radar_cc):
(TestBranch.test_manual_radar_cc):

Canonical link: https://commits.webkit.org/254577@main
